### DISCUSSION TITLE
When processing many files, I want to write each newly generated file immediately

### DISCUSSION THREAD

**Discussion by 00SS** [ImageMagick/ImageMagick#3915] [Help]
When processing multiple files, is there a way to force ImageMagick to write each new file immediately after it is generated, rather than only right at the end after all have been processed? 

Or perhaps writing the new files after a certain number have been processed - like after every 10 files?

For example - when cutting away the right half of a few hundred TIF files:

```magick image_*.tif -gravity West -crop 50%x100%+0+0 -quality 99 -set filename:0 %[fx:t*2+10001] "split\image_%[filename:0].png"```

**Comment by snibgo** [comment]
Your "magick" command will read all the tif files, then crop all the images, then write all the output files.

To do them one at a time, use "magick mogrify". See https://imagemagick.org/script/mogrify.php

**Reply by Jmuccigr** [reply]
Alternatively feed the files to magick one at a time using the shell: `for i in image_*.tif; do magick "$i" etc; done`

**Comment by snibgo** [comment]
Yes.

"magick mogrify" works one file at a time, but the syntax is different to the simple "magick".

@Jmuccigr's method has the advantage that the syntax doesn't need to change, apart from the input filename. The performance will be worse as the executable needs to be re-run for every input file, but this may not be important, and will be trivial when the images are large.

**Reply by Jmuccigr** [reply]
I'm not 100% sure of this, but perhaps running each iteration as a background process will help?

`for i in image_*.tif; do magick "$i" etc &; done`

**Comment by snibgo** [comment]
I don't know how Unix works, regarding background/foreground processes versus multiple foreground processes. But it certainly works in Windows.

The input files can be divided into groups, and each group processed at the same time. This may reduce the overall elapsed time by using CPU cores that would otherwise be unused. In the command shown, the "-crop" operation probably does use multiple cores, but reading and writing files won't.

Operating system tools can be used to determine CPU, memory and disk usage, to provide an idea of whether parallel processing would help. (On Windows, "taskmgr" can be used.) Parallel processing may use more cores, but immediately cause bottlenecks in memory or disk, limiting the performance gains.

If multiple IM processes are started, each process should be limited to a single thread. Otherwise the first process to start might use all the cores, stalling the other processes. And, of course, don't start more processes than the CPU has cores.

**Comment by 00SS** [comment]
Thanks for the suggestions. I should have mentioned this is for Windows 10 using IM v.7.1.0-Q16

I didn't realize `mogrify` worked this way, and it's good to know. Unfortunately I can't use it because the actual much longer command line also includes `-deskew` which doesn't work with `mogrify`

Though I thought the problem was files not being written one by one, I realized the main problem was the command using up too much processing power, slowing down the computer for too long. I believe I have solved it by using `-limit thread 2` :

```magick image_*.tif -limit thread 2 -gravity West -crop 50%x100%+0+0 -quality 99 -set filename:0 %[fx:t*2+10001] "split\image_%[filename:0].png"```

Thank you.
